client-v2/src/main/java/com/clickhouse/client/api/data_formats/ClickHouseBinaryFormatReader.java

```diff
@@ -0,0 +1,150 @@
+package com.clickhouse.client.api.data_formats;
+
+import com.clickhouse.data.value.ClickHouseGeoMultiPolygonValue;
+import com.clickhouse.data.value.ClickHouseGeoPointValue;
+import com.clickhouse.data.value.ClickHouseGeoPolygonValue;
+import com.clickhouse.data.value.ClickHouseGeoRingValue;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+public interface ClickHouseBinaryFormatReader {
+
+    /**
+     * Reads a single value from the stream.
+     *
+     * @param <T>
+     * @return
+     * @throws IOException
+     */
+    <T> T readValue(int colIndex) throws IOException;
+
+    /**
+     * Reads a row to an array of objects.
+     *
+     * @param colName
+     * @param <T>
+     * @return
+     * @throws IOException
+     */
+    <T> T readValue(String colName);
+
+    /**
+     * Checks if there are more rows to read.
+     *
+     * @return
+     */
+    boolean hasNext();
+
+    /**
+     * Moves cursor to the next row. Must be called before reading the first row.
+     *
+     * @return true if there are more rows to read, false otherwise
+     */
+    boolean next();
+
+    /**
+     * Copies current record to a map.
+     *
+     * @param destination - map to which the record will be copied
+     * @throws IOException
+     */
+    void copyRecord(Map<String, Object> destination) throws IOException;
+
+    String getString(String colName);
+
+    Byte getByte(String colName);

```

Comment:
Should we have also 
byte getByte(int index)

Reply:
My thought process was: 
- using index on its own requires sync with query all the time otherwise wrong column may be accessed
- so there should be a map between name and index 
- so I've just built-in this

If someone want to just read all columns in the loop - we may provide an iterator.

Reply:
if we have the case that the client asks to return the data in CSV,TSV format we have indexes